File: app/src/main/java/com/redteamobile/smart/MainActivity.java
package com.redteamobile.smart;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.core.app.ActivityCompat;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.redteamobile.smart.agent.AgentService;
import com.redteamobile.smart.recycler.ProfileAdapter;
import com.redteamobile.smart.recycler.ProfileModel;
import com.redteamobile.smart.util.SharePrefSetting;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends Activity implements View.OnClickListener {

    private static final String TAG = "MainActivity";
    private RecyclerView profileRecycler;
    private TextView enableNextTv;
    private TextView eidTv;
    private TextView imeiTv;
    private TextView appVersionTv;
    private TextView currentIccidTv;
    private TextView euiccMode;
    private TextView vuiccMode;
    private TextView enableProTv;
    private TextView finishTv;

    private AgentService agentService;
    private final byte[] eId = new byte[32];
    private final int[] eIdLength = new int[1];
    private List<ProfileModel> profileList = new ArrayList();
    private final int PERMISSON_REQUESTCODE = 110;
    private String[] needPermissions = {Manifest.permission.READ_PHONE_STATE};
    private int currentIccidIndex = -1;
    private final BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (Constant.ACTION_NOTIFY_STATE.equals(intent.getAction())) {
                showData();
                parseProfiles();
            }
        }
    };

    private ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            agentService = ((AgentService.MyBinder) service).getService();
            ProfileAdapter adapter = (ProfileAdapter) profileRecycler.getAdapter();
            if (adapter != null) {
                adapter.setAgentService(agentService);
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        checkPermmission();
        profileRecycler = findViewById(R.id.profileRecycler);
        enableNextTv = findViewById(R.id.enable_next_operationalTv);
        eidTv = findViewById(R.id.eidTv);
        imeiTv = findViewById(R.id.imeiTv);
        appVersionTv = findViewById(R.id.appVersionTv);
        currentIccidTv = findViewById(R.id.currentIccid);
        euiccMode = findViewById(R.id.euiccMode);
        vuiccMode = findViewById(R.id.vuiccMode);
        enableProTv = findViewById(R.id.enable_provisioningTv);
        finishTv = findViewById(R.id.deleteAllTv);

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        profileRecycler.setLayoutManager(linearLayoutManager);
        profileRecycler.setAdapter(new ProfileAdapter(profileList));

        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(Constant.ACTION_NOTIFY_STATE);
        LocalBroadcastManager.getInstance(this).registerReceiver(broadcastReceiver, intentFilter);

        enableNextTv.setOnClickListener(this);
        euiccMode.setOnClickListener(this);
        vuiccMode.setOnClickListener(this);
        enableProTv.setOnClickListener(this);
        finishTv.setOnClickListener(this);

        Intent intent = new Intent(this, AgentService.class);
        startService(intent);
        bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unbindService(serviceConnection);
        LocalBroadcastManager.getInstance(this).unregisterReceiver(broadcastReceiver);
    }

    private void checkPermmission() {
        ActivityCompat.requestPermissions(this,
                needPermissions, PERMISSON_REQUESTCODE);
    }

    private void updateModeUI() {
        euiccMode.setEnabled(agentService.getUiccMode() != Constant.EUICC_MODE);
        vuiccMode.setEnabled(agentService.getUiccMode() != Constant.VUICC_MODE);
    }

    private void showData() {
        if (agentService == null) {
            return;
        }
        updateModeUI();
        agentService.getEId(eId, eIdLength);
        String eid = new String(eId);
        if (TextUtils.isEmpty(eid)) {
            return;
        }
        eidTv.setText(eid);
        imeiTv.setText(agentService.getImei());
        appVersionTv.setText(BuildConfig.VERSION_NAME);
    }

    private byte[] profile = new byte[1024];
    private int[] profileLength = new int[1024];

    private void parseProfiles() {
        if (agentService == null) {
            Log.e(TAG, getString(R.string.agent_service_empty));
            return;
        }
        profileList.clear();
        agentService.getProfiles(profile, profileLength);
        byte[] jsonArray = new byte[this.profileLength[0]];
        System.arraycopy(this.profile, 0, jsonArray, 0, this.profileLength[0]);
        String json = new String(jsonArray, StandardCharsets.UTF_8);
        if (TextUtils.isEmpty(json)) {
            return;
        }
        try {
            JSONObject jsonObject = new JSONObject(json);
            JSONArray profiles = jsonObject.getJSONArray("profiles");
            for (int i = 0; i < profiles.length(); i++) {
                JSONObject jsonProfile = profiles.getJSONObject(i);
                profileList.add(new ProfileModel(jsonProfile.getString("iccid"), jsonProfile.getInt("type"),
                        jsonProfile.getInt("state")));
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
        if (profileList.size() > 0) {
            profileRecycler.getAdapter().notifyDataSetChanged();
            boolean isContains = false;
            for (int i = 0; i < profileList.size(); i++) {
                ProfileModel model = (ProfileModel) profileList.get(i);
                if (TextUtils.equals(model.getIccid(), SharePrefSetting.getCurrentIccId())) {
                    isContains = true;
                    break;
                }
            }
            ArrayList list = new ArrayList<ProfileModel>();
            for (int i = 0; i < profileList.size(); i++) {
                ProfileModel model = (ProfileModel) profileList.get(i);
                if (model.getState() == 1) {
                    currentIccidIndex = i;
                    currentIccidTv.setText(model.getIccid());
                    if (!isContains) {
                        SharePrefSetting.putCurrentIccId(model.getIccid());
                    }
                }
                if (model.getType() == 2) {
                    list.add(model);
                }
            }
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @SuppressLint("NonConstantResourceId")
    @Override
    public void onClick(View v) {
        if (profileList.size() <= 0) {
            showData();
            parseProfiles();
            return;
        }
        if (agentService == null) {
            return;
        }
        switch (v.getId()) {
            case R.id.euiccMode:
                // 需要插入EUICC卡
                agentService.setUiccMode(Constant.EUICC_MODE);
                updateModeUI();
                Toast.makeText(MainActivity.this, getString(R.string.reboot_work), Toast.LENGTH_SHORT).show();
                break;
            case R.id.vuiccMode:
                agentService.setUiccMode(Constant.VUICC_MODE);
                updateModeUI();
                Toast.makeText(MainActivity.this,  getString(R.string.reboot_work), Toast.LENGTH_SHORT).show();
                break;
            case R.id.enable_next_operationalTv:
                if (profileList.size() > 1) {
                    currentIccidIndex = (++currentIccidIndex >= profileList.size()) ? 1 : currentIccidIndex;
                }
                agentService.enableProfile(profileList.get(currentIccidIndex).getIccid());

                break;
            case R.id.enable_provisioningTv:
                for (int i = 0; i < profileList.size(); i++) {
                    ProfileModel profileModel = profileList.get(i);
                    if (profileModel.getType() == 1) {
                        if (agentService == null) {
                            return;
                        }
                        agentService.enableProfile(profileModel.getIccid());
                    }
                }
                break;
            case R.id.deleteAllTv:
                for (int i = 0; i < profileList.size(); i++) {
                    ProfileModel profileModel = profileList.get(i);
                    if (profileModel.getType() == 2) {
                        if (agentService == null) {
                            return;
                        }
                        Log.e(TAG, "onClick: " + profileModel.getIccid());
                        agentService.deleteProfile(profileModel.getIccid());
                    }
                }
                break;
        }
    }


}

File: app/src/main/java/com/redteamobile/smart/MainActivity.kt
package com.redteamobile.smart

import android.Manifest
import android.content.*
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.IBinder
import android.text.TextUtils
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.redteamobile.smart.agent.AgentService
import com.redteamobile.smart.agent.AgentService.MyBinder
import com.redteamobile.smart.agent.recycler.ProfileAdapter
import com.redteamobile.smart.agent.recycler.ProfileModel
import com.redteamobile.smart.agent.util.SharePrefSetting
import kotlinx.android.synthetic.main.activity_main.*
import org.json.JSONObject

class MainActivity : AppCompatActivity(), View.OnClickListener {

    private var agentService: AgentService? = null
    private var profile = ByteArray(1024)
    private var profileLength: IntArray = IntArray(1)

    private var eId = ByteArray(32)
    private var eIdLength: IntArray = IntArray(1)
    private var profileList = ArrayList<ProfileModel>()
    private var isNeedCheck = true
    private val PERMISSON_REQUESTCODE = 110
    private val LOCATION_CODE = 1315
    private val needPermissions = arrayOf<String>(Manifest.permission.READ_PHONE_STATE)

    private val broadcastReceiver = object : BroadcastReceiver() {

        override fun onReceive(context: Context?, intent: Intent?) {
            val action = intent?.action
            if (Constant.ACTION_NETWORK_STATE_CHANGED == action) {
                val networkState = intent?.getIntExtra(Constant.TAG_NETWORK_STATE, 0)
                if (networkState == Constant.DSI_STATE_CALL_CONNECTED) {
                    showData()
                    parseProfiles()
                }
            } else if (Constant.ACTION_SERVICE_CONNECTED == action) {
                showData()
                parseProfiles()
            }
        }

    }

    private val serviceConnection = object : ServiceConnection {

        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            agentService = (service as MyBinder).getService()
            val adapter = profileRecycler.adapter as ProfileAdapter
            adapter.setAgentService(agentService!!)
        }

        override fun onServiceDisconnected(name: ComponentName) {
        }
    }

    fun checkPermissions(permissions: Array<String>) {
        val needRequestPermissonList = findDeniedPermissions(permissions)
        if (null != needRequestPermissonList && needRequestPermissonList.size > 0) {
            ActivityCompat.requestPermissions(this,
                needRequestPermissonList.toTypedArray(), PERMISSON_REQUESTCODE)
        }
    }

    /**
     * 获取权限集中需要申请权限的列表
     * */
    fun findDeniedPermissions(permissions: Array<String>): List<String> {
        var needRequestPermissonList = ArrayList<String>()
        for (perm in permissions) {
            if (ContextCompat.checkSelfPermission(this,
                    perm) != PackageManager.PERMISSION_GRANTED || ActivityCompat.shouldShowRequestPermissionRationale(
                    this, perm)) {
                needRequestPermissonList.add(perm)
            }
        }
        return needRequestPermissonList
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, paramArrayOfInt: IntArray) {
        when (requestCode) {
            PERMISSON_REQUESTCODE -> if (!verifyPermissions(paramArrayOfInt)) {
                isNeedCheck = false
            }
            LOCATION_CODE -> {
                if (paramArrayOfInt.size > 0 && paramArrayOfInt[0] == PackageManager.PERMISSION_GRANTED) {
                    // 权限被用户同意。

                } else {
                    // 权限被用户拒绝了。
                    Toast.makeText(this@MainActivity, "定位权限被禁止，相关地图功能无法使用！", Toast.LENGTH_LONG).show()
                }

            }
        }

    }

    /**
     * 检测是否说有的权限都已经授权
     */
    fun verifyPermissions(grantResults: IntArray): Boolean {
        for (result in grantResults) {
            if (result != PackageManager.PERMISSION_GRANTED) {
                return false
            }
        }
        return true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        var linearLayoutManager = LinearLayoutManager(this)
        linearLayoutManager.orientation = LinearLayoutManager.VERTICAL
        profileRecycler.layoutManager = linearLayoutManager
        profileRecycler.adapter = ProfileAdapter(profileList)

        var intentFilter = IntentFilter()
        intentFilter.addAction(Constant.ACTION_NETWORK_STATE_CHANGED)
        intentFilter.addAction(Constant.ACTION_SERVICE_CONNECTED)
        LocalBroadcastManager.getInstance(this).registerReceiver(broadcastReceiver, intentFilter)

        enable_next_operationalTv.setOnClickListener(this)

        val intent = Intent(this, AgentService::class.java)
        startService(intent)
        bindService(intent, serviceConnection, Context.BIND_AUTO_CREATE)
        if (isNeedCheck) {
            checkPermissions(needPermissions);
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        unbindService(serviceConnection)
        LocalBroadcastManager.getInstance(this).unregisterReceiver(broadcastReceiver)
    }


    fun showData() {
        if (agentService == null) {
            return
        }

        agentService?.getEId(eId, eIdLength)
        var eid = String(eId)
        if (eid == null || eid.isEmpty()) {
            return
        }
        eidTv.text = eid
        imeiTv.text = agentService?.getImei()
        appVersionTv.text = BuildConfig.VERSION_NAME
    }

    private fun parseProfiles() {
        if (agentService == null) {
            return
        }
        profileList.clear()
        agentService?.getProfiles(profile, profileLength)
        var jsonArray = ByteArray(profileLength[0])
        System.arraycopy(profile, 0, jsonArray, 0, profileLength[0])
        val json = String(jsonArray)

        if (json == null || json.isEmpty()) {
            return
        }
        var jsonObject = JSONObject(json)
        val profiles = jsonObject.getJSONArray("profiles")
        for (i in 0 until profiles.length()) {
            var profile = profiles.getJSONObject(i)
            profileList.add(
                ProfileModel(
                    profile.getString("iccid"),
                    profile.getInt("type"),
                    profile.getInt("state")
                )
            )
        }
        if (profileList.size > 0) {
//            profileRecycler.adapter = agentService?.let { ProfileAdapter(profileList, it) }
            profileRecycler.adapter?.notifyDataSetChanged()
            var isContains = false
            for (profile in profileList) {
                if (TextUtils.equals(profile.iccid, SharePrefSetting.getCurrentIccId())) {
                    isContains = true
                    break
                }
            }
            var list = ArrayList<ProfileModel>()
            for (profile in profileList) {
                if (profile.state == 1) {
                    currentIccid.text = profile.iccid
                    if (!isContains) {
                        SharePrefSetting.putCurrentIccId(profile.iccid)
                    }
                }

                if (profile.type == 2) {
                    list.add(profile)
                }
            }

        }
    }

    override fun onClick(v: View?) {
        if (v != null) {
            if (profileList.size <= 0) {
                return
            }
            when (v.id) {
                startService.id and enable_next_operationalTv.id -> {
                    for (profile in profileList) {
                        if (profile.type == 2) {
                            if (profile.state == 0) {
                                if (agentService == null) {
                                    return
                                }
                                agentService!!.enableProfile(profile.iccid)
                            }
                        }
                    }
                }
                stopService.id -> {
                    for (profile in profileList) {
                        if (profile.type == 2) {
                            if (profile.state == 1) {
                                if (agentService == null) {
                                    return
                                }
                                agentService!!.disableProfile(profile.iccid)
                            }
                        }
                    }
                }
                enable_provisioningTv.id -> {
                    for (profile in profileList) {
                        if (profile.type == 1) {
                            if (agentService == null) {
                                return
                            }
                            agentService!!.enableProfile(profile.iccid)
                        }
                    }
                }
                qa_finishTv.id -> {
                    for (profile in profileList) {
                        if (profile.type == 2) {
                            if (agentService == null) {
                                return
                            }
                            agentService!!.deleteProfile(profile.iccid)
                        }
                    }
                }
            }

        }
    }

}
